### ISSUE TITLE
Angular and Clarity upgrade - documentation for local execution

### ISSUE THREAD

**Issue by jhiemer** [vmware-samples/vcd-ext-samples#9]
**Is your feature request related to a problem? Please describe.**
I was checking out the project to create an extension for vCD. Having worked with Clarity before I was surprised that the sample project is using Angular 4.4, while the current version of Clarity recommends Angular 6.

Next thing: as we plan to develop a bit more complicated plugin, it would be nice to have a default setup to run the app locally. 

`npm start` results in:

```
npm start
npm ERR! missing script: start

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/jhiemer/.npm/_logs/2019-09-23T18_43_22_927Z-debug.log
```
Last question: why is the sample using `npm` and not `yarn` although it is recommended by the Angular team?

**Comment by jmoroski** [comment]
> I was surprised that the sample project is using Angular 4.4, while the current version of Clarity recommends Angular 6.

The sample project is currently using Angular 4.x because it is using Clarity 0.10.x, and the reason for this is backward compatibility. The plugins that you author run within the vCloud Director's UI, and they must be compatible with the versions of libraries for RxJS, Angular, and Clarity that vCD is using.

The sample project is backward compatible to version 9.1.0.2 of vCD, and that is the reason for the use of these library versions.

We are currently looking to create different samples that use newer Angular/Clarity versions, at the expense of vCD backward compatibility, for plugin authors who want to leverage newer libraries, or don't have as large of an interop matrix to support.

>  it would be nice to have a default setup to run the app locally

Agreed, but in this case plugins don't run without a vCD installation. There are a few Python scripts provided to push plugins to a vCD instance (which _could_ be running locally). These can currently be put behind publish/deploy/start scripts, and we're looking at a pure JS solution to incorporate the capabilities into the development process without needing Python.

**Comment by jhiemer** [comment]
@jmoroski thanks for the quick response. Regarding your statements:

> We are currently looking to create different samples that use newer Angular/Clarity versions, at the expense of vCD backward compatibility, for plugin authors who want to leverage newer libraries, or don't have as large of an interop matrix to support.
Do you have a timeline for it?

>These can currently be put behind publish/deploy/start scripts, and we're looking at a pure JS solution to incorporate the capabilities into the development process without needing Python.
What would be the way to go here? If I understand it correctly regarding the UI setup a plugin just provides the left column menu (if needed) and the content area. The upper navigation and the bottom section are vCD native. If an author develops a plugin which is talking to external APIs anyway I don't see too much benefit of deploying the application each time I want to run/debug it. If this is the case, I am loosing all of the reload functionality within `ng serve` etc...and really makes it a huge pain to develop more complex extensions.

**Comment by HynesIP** [comment]
@jhiemer thanks for raising this issue. @jmoroski thanks for the quick responses. This is what I was leading to with my questions at https://github.com/vmware/vcd-ext-sdk/issues/92 this weekend.

I've spent some time trying to either add >ng serve to this project or import this project into a newer Clarity sandbox. Nothing has worked for me. Seems Angular 4.4.4 isn't directly associated with an Angulsr CLI version that I can find. When I try to install Angular manually starting with core 4.4.4, oddly all the related Angular parts starts asking for core 4.4.7 as the requirement. Very frustrating.

Standing by to try whatever you come up with.

**Comment by jhiemer** [comment]
@HynesIP good to see that I am not the only one running into those issues. I was already starting to question myself. 

@jmoroski I ran into another question today. In the docs here: https://github.com/vmware/vcd-ext-sdk/blob/master/ui/vcd-plugin-seed/README.md I read the following: 
>The data for each plugin is loaded from the server in isolation from any other plugins, and in isolation from vCD itself. In order for the plugin to be able to load its assets, it needs to be able to specify the root URL of those assets, which is created and managed by vCD.

Which made me think, that I don't need to care about the compatibility between the vCD Angular and Clarity version and the version of the plugin itself. But if I understand it correctly, this assumption is wrong.

A further question I have: when I look at the structure of the seed plugin, with the placement of the modules and components etc. this does not look the way it is recommended and generate by `ng g component/module`, especially the `main` folder compared to the `app` folder in a "normal" Angular app. 

Looking forward to your feedback,
Johannes

**Comment by jmoroski** [comment]
@jhiemer 
> I don't need to care about the compatibility between the vCD Angular and Clarity version and the version of the plugin itself. But if I understand it correctly, this assumption is wrong.

It is the data that is scoped to the plugin along with any libraries that aren't marked as external in the webpack configuration. Clarity is marked as external and not bundled with the plugin because Clarity registers itself with the global namespace (or it did as recently as 0.13.x) which means that it was impossible to bundle a version of Clarity with a plugin and not have it collide with the version of Clarity that the vCD UI was using. So the code that loads the plugins bootstraps them into an environment where Angular, RxJS, and Clarity are already available. That is the source of the limitation on specific library versions.

An advantage to the current sharing of fundamental libraries (despite its other limitations) is the plugin size and bootstrapping time/bandwidth. Angular and Clarity are not small, even with tree shaking, and currently each plugin doesn't incur a heavy cost to load. We can also more easily share theme and branding between Clarity components of the same version which is an important consideration for vCD and its goal of being a whitelabel product.

Both of these requirements are what shaped the decision to not just embed plugins as iframes (though you are still free to write a plugin that is an iframe), but to provide a more rich and native integration.

We are looking at ways to mitigate the library incompatibilities. I don't have anything useful yet, but I'll be sure to keep the community posted.

>  Is it possible to stick to the correct setup for an Angular app?

Yes, I'm sure it is. This was originally all laid out prior to the Angular CLI really being as prevalent. It has migrated from its early implementation of yarn and rollupjs to npm and webpack. It can probably migrate even further and have a more standard structure, possibly even schematics to generate plugin components rather than a plugin seed. Contributions welcome! This topic is also a good segue into observations also shared by @HynesIP and it looks like there are some opportunities to support different plugin "starting points"–either seed projects or schematics–tailored toward the level of vCD backward compatibility that is needed, and some of the ng tooling would definitely work with newer versions of vCD in that case. Maybe not `ng serve` but certainly some of the other structure and functionality.

**Comment by jhiemer** [comment]
@jmoroski thanks again for the response. Regarding the development experience: do you think it would be possible to have a wrapper of the basic menu structure of vCD provided by VMware in a seed project?

We are currently setting up vCD 10 to have the latest version running. As I could not find anything in the release note section. Do you know, which Angular/Clarity version the latest vCD UI is running on? I am currently in the process of setting up a project with the Angular CLI based module and component generation to test it with vCD  10.

**Comment by HynesIP** [comment]
Yes, @jmoroski thank you.

**Comment by jmoroski** [comment]
At this time I don't have a good answer for the more iterative development process. Yes, it is probably possible to create a wrapper of the basic vCD UI shell. It would have to be structured to bootstrap plugins the way the vCD UI does which is a proprietary blend of SystemJS, clever Clarity shims, an extension registry, and magic. That is a non-trivial amount of work, and something that needs ongoing maintenance as the process evolves from release to release.

That effort _only_ buys some ease of development for top level extensibility (adding a new primary navigation item). Datacenter Overview extensibility needs the actual Datacenters navigation in place. VM and vApp action menu extensibility need the Datacenters navigation also, along with the Datacenter specific routes, and actual vApps and VMs. It quickly snowballs into something that is pretty much a duplicate copy of the vCD UI.

I feel this pain myself since I'm usually the one building plugins to test SDK changes. I think any tenable solution at this point needs to involve an actual vCD instance, but maybe there is a way to get it to reference a watched build directory from a dev machine rather than demanding a full ZIP distribution. If there's a way to short circuit that process a bit it could tighten up the development cycle.

**Comment by jhiemer** [comment]
@jmoroski I checked out the latest release of vCD 10. I see it is based on Angular 6.1.10
![image](https://user-images.githubusercontent.com/1557826/65814434-de997180-e1e1-11e9-939f-7e7014c8c847.png). But I could not find the Clarity version. Do you know it?

>It would have to be structured to bootstrap plugins the way the vCD UI does which is a proprietary blend of SystemJS, clever Clarity shims, an extension registry, and magic.

I guess there is plan of releasing blend of the plugin system for the vCD as part of this repo?

> Datacenter Overview extensibility needs the actual Datacenters navigation in place. VM and vApp action menu extensibility need the Datacenters navigation also, along with the Datacenter specific routes, and actual vApps and VMs. It quickly snowballs into something that is pretty much a duplicate copy of the vCD UI.

Understood. But on the other hand I would expect one could get this up and running with some mocked objects, or by enabling CORS requests on a development vCD instance. This would then just call the normal endpoints for `/cloudapi/1.0.0` and `/api` and one could start developing locally, at least this is what we do in other Angular based projects.

I will starting working on a seed based on the vCD 10 and look into the development experience, but currently I plan to develop the modules and components in a parallel project and when implemented and tested, migrate it into the seed.

**Comment by jmoroski** [comment]
Clarity version for vCD 10 is 0.13.4.

> I plan to develop the modules and components in a parallel project and when implemented and tested, migrate it into the seed.

I've teased this idea in the past. We've tried to ensure that the project is a normal Angular project as far as modules, components, routes, etc. We've said that this should be useful for someone who both wants to build a standalone portal, but still make portions of it available within vCD. So I see no reason that _wouldn't_ work, I just haven't yet gotten around to it myself. In fact, I don't even see why  it would then need to be copied into a separate project. If it's the same generated bundle.js, and some additional assets and a manifest for the plugin, that could just be an extra dist script in your package.json.

**Comment by carceneaux** [comment]
+1 for this. As it was previously mentioned, this would allow for automated testing (and testing locally) making it easier to develop more complex plugins using the vcd-ext-sdk.

**Comment by misterGF** [comment]
+1. This project really needs a way to help with local development.

**Comment by carceneaux** [comment]
As @jmoroski previously mentioned, here's my workaround for this today.

I use the [Plugin Manager](https://github.com/vmware-samples/vcd-ext-samples/tree/plugin-manager) hosted in this repo. To make the dev flow as easy as possible, I also added the following lines to the `scripts` section in my `package.json`:

```json
"deploy": "python ./manage_plugin.py register",
"remove": "python ./manage_plugin.py unregister",
"list": "python ./manage_plugin.py list"
```

This allows me to test the plugin with the following one-liner:

`npm run build; npm run deploy`

Those familiar with early versions of the `vcd-plugin-seed` will recognize this process.

**Comment by jmoroski** [comment]
The new developer-container/9.7-10.0 branch contains my first pass at speeding up the dev lifecycle. There is an overview of it here:
https://blogs.vmware.com/cloudprovider/2020/02/cloud-director-plugin-dev-env.html

The README in the cloud-director-container/ folder of the above branch contains a (very) basic description of adding a plugin. I hope to have another blog post published dedicated to building out a plugin in this environment within the next week.

Does this start to address people's needs?

@jhiemer @HynesIP @carceneaux

**Comment by carceneaux** [comment]
Great job @jmoroski!

From what I've seen thus far, it does. As I recently finished my dev cycle on a plugin and I was OOO last week, I haven't been able to check this out yet. Give me the rest of the week and I'll stand it up in my environment and provide further feedback.

**Comment by misterGF** [comment]
Thanks @jmoroski  I'll be trying it out soon but upon first look, it looks like a great improvement.

**Comment by misterGF** [comment]
> Thanks @jmoroski I'll be trying it out soon but upon first look, it looks like a great improvement.

I was able to get this going with little résistance! Thanks for that. The only thing I had to change was the version number passed in the headers. You have version 34.0 but I'm using 33.0 (vcd v10). This can be found in app-config.service.ts line 41.

**Comment by jmoroski** [comment]
> You have version 34.0 but I'm using 33.0 (vcd v10). 

That's great! You're certainly past starting the container and toying with branding to notice this. I'll push a correction tomorrow.

**Comment by carceneaux** [comment]
@jmoroski Set this up in my lab today following your [instructions](https://github.com/vmware-samples/vcd-ext-samples/blob/developer-container/9.7-10.0/cloud-director-container/README.md).

Things went pretty smooth. As I haven't used `lazyModules` previously, it took about a minute of googling to figure out that I needed to put this in the `options` section. I recommend you add some verbiage there for clarity.

You mentioned the process for adding plugins was to `copy them by hand`. I recommend using a symbolic link. This allows the two projects to be separate and forgoes any copy pasta errors. They are supported on both Mac & Windows so this shouldn't be an issue.

Does this help the dev process? **Immensely!** Thanks! Once vCD 9.5 support is done, I'll migrate my current dev process to this. Until then, I'm tied to Angular 4.4 for backwards compatibility.

Sorry this took so long. Thanks again!

**Comment by HynesIP** [comment]
@jmoroski Just what I've been waiting for. Seems so simple after seeing under the hood.

Thank you.

**Comment by jmoroski** [comment]
Well... I mean... It wasn't _that_ simple 😆

**Comment by HynesIP** [comment]
Just means you did a great job. ;-)

Any ETA on the blog post for building the plugin. Should I just be able to port over the webpack from the original project?

**Comment by shubhamkr-cohesity** [comment]
@jmoroski Can you also update the dependency matrix for vCD 10.1?

Also, is the seed app qualified for vCD 10.1? 
I am having some routing issues. (Already reported)

**Comment by kbidve** [comment]
@jmoroski 

Unable to install the dependency, facing following issue:

npm WARN prepare removing existing node_modules/ before installation

> fsevents@1.2.11 install 
> node-gyp rebuild


>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild )
gyp WARN install got an error, rolling back install
gyp ERR! configure error
gyp ERR! stack Error: self signed certificate in certificate chain
gyp ERR! stack     at TLSSocket.onConnectSecure (_tls_wrap.js:1474:34)
gyp ERR! stack     at TLSSocket.emit (events.js:310:20)
gyp ERR! stack     at TLSSocket._finishInit (_tls_wrap.js:917:8)
gyp ERR! stack     at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:687:12)
gyp ERR! System Windows_NT 10.0.18363
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\kailas.bidve\Documents\Mcloude\UI_Code\Dev_Enviornment\vcd-dev-env\node_modules\fsevents
gyp ERR! node -v v12.16.3
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok
added 264 packages in 7.057s

>npm start
npm ERR! missing script: start

npm ERR! A complete log of this run can be found in:
npm ERR!

**Comment by kouassives** [comment]
> The new developer-container/9.7-10.0 branch contains my first pass at speeding up the dev lifecycle. There is an overview of it here: https://blogs.vmware.com/cloudprovider/2020/02/cloud-director-plugin-dev-env.html
> 
> The README in the cloud-director-container/ folder of the above branch contains a (very) basic description of adding a plugin. I hope to have another blog post published dedicated to building out a plugin in this environment within the next week.
> 
> Does this start to address people's needs?
> 
> @jhiemer @HynesIP @carceneaux

Hello @jmoroski,

I have a problem while using the container. After cloning the developer-container / 9.7-10.0. I followed the Readme.md in the cloud-director-container folder. I therefore cloned the plugin-seed project in the src / plugins directory of cloud-director-container and add the necessary modifications in angular.json and index.ts in src / plugins

But I have a problem to load the plugins as you can see in the screenshot. 
**ERROR Error: Uncaught (in promise): Error: No NgModule metadata found for 'SimplePluginModule'.
Error: No NgModule metadata found for 'SimplePluginModule'.**

Please can you help me fix it?

<img width="418" alt="image" src="https://user-images.githubusercontent.com/34067783/144708971-275440a9-673e-4217-b4cc-ea65886bd1c4.png">

**Comment by nvladimirovi** [comment]
@kouassives Please refer to our documentation https://vmware.github.io/vcd-ext-sdk/
